File: android/app/src/main/java/com/classroom/cast/CastClient.kt
package com.classroom.cast

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okio.ByteString.Companion.toByteString
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONObject
import java.security.SecureRandom
import java.security.cert.X509Certificate
import java.util.concurrent.TimeUnit
import javax.net.ssl.SSLContext
import javax.net.ssl.X509TrustManager

sealed class CastEvent {
    data class Connected(val serverInfo: JSONObject) : CastEvent()
    data class Error(val message: String) : CastEvent()
    data object Disconnected : CastEvent()
    data class Message(val text: String) : CastEvent()
}

class CastClient(private val host: String, private val port: Int, private val useSsl: Boolean = false) {

    private val scheme get() = if (useSsl) "https" else "http"
    private val wsScheme get() = if (useSsl) "wss" else "ws"
    private val baseUrl get() = "$scheme://$host:$port"

    private val client = buildClient()

    private var ws: WebSocket? = null
    private val _events = Channel<CastEvent>(Channel.BUFFERED)
    val events: Flow<CastEvent> = _events.receiveAsFlow()

    private fun buildClient(): OkHttpClient {
        val builder = OkHttpClient.Builder()
            .connectTimeout(5, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .readTimeout(0, TimeUnit.MILLISECONDS) // no read timeout for WS

        if (useSsl) {
            // Trust all certificates (self-signed cert on the classroom server)
            val trustAll = object : X509TrustManager {
                override fun checkClientTrusted(certs: Array<out X509Certificate>?, authType: String?) {}
                override fun checkServerTrusted(certs: Array<out X509Certificate>?, authType: String?) {}
                override fun getAcceptedIssuers(): Array<X509Certificate> = arrayOf()
            }
            try {
                val sslContext = SSLContext.getInstance("TLS")
                sslContext.init(null, arrayOf(trustAll), SecureRandom())
                builder.sslSocketFactory(sslContext.socketFactory, trustAll)
                builder.hostnameVerifier { _, _ -> true }
            } catch (_: Exception) {}
        }

        return builder.build()
    }

    fun connectWs() {
        val url = "$wsScheme://$host:$port/ws"
        val request = Request.Builder()
            .url(url)
            .build()

        client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(ws: WebSocket, response: Response) {
                this@CastClient.ws = ws
            }

            override fun onMessage(ws: WebSocket, text: String) {
                _events.trySend(CastEvent.Message(text))
            }

            override fun onFailure(ws: WebSocket, t: Throwable, response: Response?) {
                _events.trySend(CastEvent.Error(t.message ?: "WebSocket error"))
            }

            override fun onClosed(ws: WebSocket, code: Int, reason: String) {
                _events.trySend(CastEvent.Disconnected)
            }
        })
    }

    fun sendFrame(jpegData: ByteArray) {
        ws?.send(jpegData.toByteString())
    }

    fun sendCastStart() {
        ws?.send("CAST_START")
    }

    fun sendCastStop() {
        ws?.send("CAST_STOP")
    }

    fun closeWs() {
        ws?.close(1000, "User stopped")
        ws = null
    }

    suspend fun fetchServerInfo(): JSONObject? {
        return try {
            withContext(Dispatchers.IO) {
                val request = Request.Builder()
                    .url("$baseUrl/api/info")
                    .build()
                val response = client.newCall(request).execute()
                JSONObject(response.body?.string() ?: return@withContext null)
            }
        } catch (e: Exception) {
            null
        }
    }

    suspend fun uploadPhoto(data: ByteArray, filename: String): Boolean {
        return try {
            withContext(Dispatchers.IO) {
                val body = MultipartBody.Builder()
                    .setType(MultipartBody.FORM)
                    .addFormDataPart("photo", filename, data.toRequestBody("image/jpeg".toMediaType()))
                    .build()
                val request = Request.Builder()
                    .url("$baseUrl/api/upload")
                    .post(body)
                    .build()
                val response = client.newCall(request).execute()
                val json = JSONObject(response.body?.string() ?: "{}")
                json.optBoolean("ok", false)
            }
        } catch (e: Exception) {
            false
        }
    }

    suspend fun pushUrl(url: String): Boolean {
        return try {
            withContext(Dispatchers.IO) {
                val json = JSONObject().apply { put("url", url) }
                val body = json.toString().toRequestBody("application/json".toMediaType())
                val request = Request.Builder()
                    .url("$baseUrl/api/pushurl")
                    .post(body)
                    .build()
                val response = client.newCall(request).execute()
                val respJson = JSONObject(response.body?.string() ?: "{}")
                respJson.optBoolean("ok", false)
            }
        } catch (e: Exception) {
            false
        }
    }

    fun disconnect() {
        closeWs()
    }
}

File: ios/ClassroomCast/Services/CastClient.swift
import Foundation

// MARK: - Cast Event
enum CastEvent {
    case connected
    case disconnected
    case error(String)
    case message(String)
}

// MARK: - Cast Client
class CastClient: NSObject, URLSessionWebSocketDelegate {
    let host: String
    let port: Int

    private var ws: URLSessionWebSocketTask?
    private var session: URLSession!
    private var onEvent: ((CastEvent) -> Void)?

    init(host: String, port: Int) {
        self.host = host
        self.port = port
        super.init()
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 10
        config.timeoutIntervalForResource = 300
        session = URLSession(configuration: config, delegate: self, delegateQueue: nil)
    }

    // MARK: - WebSocket Connection
    func connectWs(onEvent: @escaping (CastEvent) -> Void) {
        self.onEvent = onEvent
        let url = URL(string: "ws://\(host):\(port)/ws")!
        ws = session.webSocketTask(with: url)
        ws?.resume()
        receiveMessage()
    }

    private func receiveMessage() {
        ws?.receive { [weak self] result in
            switch result {
            case .success(let message):
                switch message {
                case .string(let text):
                    self?.onEvent?(.message(text))
                case .data:
                    break
                @unknown default:
                    break
                }
                self?.receiveMessage()
            case .failure(let error):
                self?.onEvent?(.error(error.localizedDescription))
            }
        }
    }

    func sendFrame(_ data: Data) {
        ws?.send(.data(data)) { _ in }
    }

    func sendText(_ text: String) {
        ws?.send(.string(text)) { _ in }
    }

    func sendCastStart() { sendText("CAST_START") }
    func sendCastStop() { sendText("CAST_STOP") }

    func closeWs() {
        ws?.cancel(with: .normalClosure, reason: nil)
        ws = nil
    }

    // MARK: - HTTP API
    func fetchServerInfo() async -> [String: Any]? {
        let url = URL(string: "http://\(host):\(port)/api/info")!
        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            return try JSONSerialization.jsonObject(with: data) as? [String: Any]
        } catch {
            return nil
        }
    }

    func uploadPhoto(_ data: Data, filename: String) async -> Bool {
        let url = URL(string: "http://\(host):\(port)/api/upload")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"

        let boundary = UUID().uuidString
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"photo\"; filename=\"\(filename)\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: image/jpeg\r\n\r\n".data(using: .utf8)!)
        body.append(data)
        body.append("\r\n--\(boundary)--\r\n".data(using: .utf8)!)
        request.httpBody = body

        do {
            let (respData, _) = try await URLSession.shared.data(for: request)
            if let json = try JSONSerialization.jsonObject(with: respData) as? [String: Any] {
                return json["ok"] as? Bool ?? false
            }
        } catch {}
        return false
    }

    func pushUrl(_ urlStr: String) async -> Bool {
        let url = URL(string: "http://\(host):\(port)/api/pushurl")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let body = ["url": urlStr]
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        do {
            let (data, _) = try await URLSession.shared.data(for: request)
            if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
                return json["ok"] as? Bool ?? false
            }
        } catch {}
        return false
    }

    // MARK: - URLSessionWebSocketDelegate
    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didOpenWithProtocol: String?) {
        DispatchQueue.main.async { self.onEvent?(.connected) }
    }

    func urlSession(_ session: URLSession, webSocketTask: URLSessionWebSocketTask, didCloseWith closeCode: URLSessionWebSocketTask.CloseCode, reason: Data?) {
        DispatchQueue.main.async { self.onEvent?(.disconnected) }
    }

    deinit {
        closeWs()
    }
}
